packages/cli/src/utils/run-all-tests.utils.ts

```diff
@@ -39,3 +56,69 @@ export const getTestsToRun: (options: {
           cached.title === title
       )
   );
+
+  const testCasesMissingBaseReplayId = uncachedTestCases.filter(
+    (testCase) => testCase.baseReplayId == null
+  );
+  const testCasesWithBaseReplayId = uncachedTestCases.flatMap(
+    (testCase): TestCase[] =>
+      // Note: explictly setting the baseReplayId is required for typescript to be happy with types
+      testCase.baseReplayId == null
+        ? []
+        : [{ ...testCase, baseReplayId: testCase.baseReplayId }]
+  );
+
+  if (testCasesMissingBaseReplayId.length === 0) {
+    return testCasesWithBaseReplayId;
+  }
+
+  if (baseCommitSha == null) {
+    const namesOfInvalidTests = testCasesMissingBaseReplayId
+      .map((test) => `"${test.title}"`)
+      .join(", ");
+    throw new Error(
+      `The following test cases are missing a baseReplayId: ${namesOfInvalidTests}.` +
+        " Please either run with the --baseCommitSha option, or add a baseReplayId to all test cases."
+    );
+  }
+
+  const baseReplayIdBySessionId = await getBaseReplayIdsBySessionId({
+    client,
+    baseCommitSha,
+  });
+
+  return uncachedTestCases.flatMap((test) => {
+    if (test.baseReplayId != null) {
+      // Note: explictly setting the baseReplayId is required for typescript to be happy with types
+      return [{ ...test, baseReplayId: test.baseReplayId }];
+    }
+    const baseReplayId = baseReplayIdBySessionId[test.sessionId];
+    if (baseReplayId == null) {
+      const logger = log.getLogger(METICULOUS_LOGGER_NAME);
+      logger.warn(
+        `Skipping comparisons for test "${test.title}" since no result to compare against stored for base commit ${baseCommitSha}`
+      );
+      return test;
+    }
+    return [{ ...test, baseReplayId }];
+  });
+};
+
+const getBaseReplayIdsBySessionId = async ({
+  client,
+  baseCommitSha,
+}: {
+  client: AxiosInstance;
+  baseCommitSha: string;
+}): Promise<Record<string, string>> => {
+  const baseReplays = await getCachedTestRunResults({
```

Comment:
This won't work for any users that have multiple run-all-tests commands setup on their CI, or are occasionally running run-all-tests locally. So I'll do a follow up PR after this to update the `/cache` endpoint to take some additional parameters. That could be a new param you pass via the CLI that uniquely identifies the test run comparison set, and the user provides, so that we can compare like for like (`--comparisonSet "main-ci-screenshots"`). Or maybe we pass to the /cache endpoint some information that helps it narrow down which test run to return, when there are multiple test runs for a commit, and it returns the closest matching one. Some possible info we could match on, each with their own tradeoffs:

 - Whether it's running on CI and the workflow name (https://docs.github.com/en/actions/learn-github-actions/contexts#github-context)
 - The meticulous.json file path within the repo
 - The session ids, and the various parameters for the replay

Maybe the best solution would be:

 - We store whether a test run was run on CI or not.
 - The cache endpoint returns the most recent test run on CI if there are multiple test runs for the commit.
 - If users want to run multiple different sets of tests on CI simultaneously then they must provide a unique `--comparisonSet` or `--testSetId` parameter for each separate call to run-all-tests. Potentially this testSetId would default to the meticulous.json path within the repo.

Reply:
Why not make an endpoint which allows to resolve: `(sessionId, commit) => latest simulation id`?

Reply:
Because there may be multiple replays for a session on the same commit run in different environments (localhost vs server; different settings - e.g. on our CI we run with/without --skipPauses etc.; eventually customers may want to run tests with dark mode on vs dark mode off etc.)

Reply:
Yep proposal would be (1) start out with this PR as is, so it risks matching wrong screenshot set if you also run run-all-tests locally on commits on your main branch (b) then update to doing the CI vs non-CI thing, and then (c) at some time later when we need it extend `test-runs/cache`/`test-runs/latest-run` to also support passing in the testSuiteId parameter.